solutions/izho-17-LongestBeautifulSequence.mdx

```diff
@@ -0,0 +1,195 @@
+---
+id: izho-17-LongestBeautifulSequence
+source: IZhO 2017
+title: 2017 - Longest beautiful sequence
+author: Shijie Ren
+---
+
+<Spoiler title="Hint 1">
+
+Both $\mathcal{O}(N^2)$ and $\mathcal{O}(N\max{a[i]})$ algorithms are not hard
+to come up with, but the latter one is more enlightening.
+
+</Spoiler>
+
+<Spoiler title="Hint 2">
+
+You may find two ways of $\mathcal{O}(N\max{a[i]})$ DP.
+
+One is kind of preprocessing previous answers, traverse all the situations of
+the previous number, $\mathcal{O}(\max{a[i]})$ to query current answer in DP
+array and $\mathcal{O}(1)$ to update.
+
+The other is kind of predicting next answers, traverse all the possible outcomes
+of the next number, $\mathcal{O}(1)$ to query current answer in DP array and
+$\mathcal{O}(\max{a[i]})$ to update, but needs a bit more memory.

```

Comment:
what's the DP state? how exactly do you transition?

Reply:
their details are in the explanation

Reply:
move it here so it flows better then